src/BSON/Int64.c

```diff
@@ -0,0 +1,377 @@
+
+	return &intern->std;
+#else
+	{
+		zend_object_value retval;
+		retval.handle   = zend_objects_store_put(intern, (zend_objects_store_dtor_t) zend_objects_destroy_object, php_phongo_int64_free_object, NULL TSRMLS_CC);
+		retval.handlers = &php_phongo_handler_int64;
+
+		return retval;
+	}
+#endif
+} /* }}} */
+
+static int php_phongo_int64_compare_objects(zval* o1, zval* o2 TSRMLS_DC) /* {{{ */
+{
+	php_phongo_int64_t *intern1, *intern2;
+
+	intern1 = Z_INT64_OBJ_P(o1);
+	intern2 = Z_INT64_OBJ_P(o2);
+
+	if (intern1->integer != intern2->integer) {
+		return intern1->integer < intern2->integer ? -1 : 1;
+	}
+
+	return 0;
+} /* }}} */
+
+static HashTable* php_phongo_int64_get_gc(zval* object, phongo_get_gc_table table, int* n TSRMLS_DC) /* {{{ */
+{
+	*table = NULL;
+	*n     = 0;
+
+	return Z_INT64_OBJ_P(object)->properties;
+} /* }}} */
+
+HashTable* php_phongo_int64_get_properties_hash(zval* object, bool is_debug TSRMLS_DC) /* {{{ */
+{
+	php_phongo_int64_t* intern;
+	HashTable*          props;
+	char                s_integer[24];
+	int                 s_integer_len;
+
+	intern = Z_INT64_OBJ_P(object);
+
+	PHONGO_GET_PROPERTY_HASH_INIT_PROPS(is_debug, intern, props, 2);
+
+	if (!intern->initialized) {
+		return props;
+	}
+
+	s_integer_len = snprintf(s_integer, sizeof(s_integer), "%" PRId64, intern->integer);
+
+#if PHP_VERSION_ID >= 70000
+	{
+		zval value;
+
+		ZVAL_STRINGL(&value, s_integer, s_integer_len);
+		zend_hash_str_update(props, "int", sizeof("int") - 1, &value);
+	}
+#else
+	{
+		zval* value;
+
+		MAKE_STD_ZVAL(value);
+		ZVAL_STRINGL(value, s_integer, s_integer_len, 1);
+		zend_hash_update(props, "int", sizeof("int"), &value, sizeof(value), NULL);
+	}
+#endif
+
+	return props;
+} /* }}} */
+
+static HashTable* php_phongo_int64_get_debug_info(zval* object, int* is_temp TSRMLS_DC) /* {{{ */
+{
+	*is_temp = 1;
+	return php_phongo_int64_get_properties_hash(object, true TSRMLS_CC);
+} /* }}} */
+
+static HashTable* php_phongo_int64_get_properties(zval* object TSRMLS_DC) /* {{{ */
+{
+	return php_phongo_int64_get_properties_hash(object, false TSRMLS_CC);
+} /* }}} */
+/* }}} */
+
+void php_phongo_int64_init_ce(INIT_FUNC_ARGS) /* {{{ */
+{
+	zend_class_entry ce;
+
+	INIT_NS_CLASS_ENTRY(ce, "MongoDB\\BSON", "Int64", php_phongo_int64_me);
+	php_phongo_int64_ce                = zend_register_internal_class(&ce TSRMLS_CC);
+	php_phongo_int64_ce->create_object = php_phongo_int64_create_object;
+	PHONGO_CE_FINAL(php_phongo_int64_ce);
+
+	zend_class_implements(php_phongo_int64_ce TSRMLS_CC, 1, php_phongo_json_serializable_ce);
+	zend_class_implements(php_phongo_int64_ce TSRMLS_CC, 1, php_phongo_type_ce);
+	zend_class_implements(php_phongo_int64_ce TSRMLS_CC, 1, zend_ce_serializable);
+
+	memcpy(&php_phongo_handler_int64, phongo_get_std_object_handlers(), sizeof(zend_object_handlers));
+	php_phongo_handler_int64.compare_objects = php_phongo_int64_compare_objects;
+	php_phongo_handler_int64.get_debug_info  = php_phongo_int64_get_debug_info;
```

Comment:
Would it make sense to add a cast handler? Like GMP does: https://github.com/php/php-src/blob/master/ext/gmp/gmp.c#L412

Reply:
That might alleviate some concerns I had with https://github.com/mongodb/mongo-php-library/pull/542/files/677838f5ca8030942686ec24e1e6b5407ffdeac4..b75386e36c6e824fd045c7ab0bcb5d65efba4b65#r197290561, although I'm not sure it will make much difference in practice.

This object will only be created if the integer's size exceeds the range of PHP's native integer type. In that case, attempting to cast it to `(integer)` is going to certainly result in a loss of data. I think the only safe approach for users is to cast as a string and they can then use that value to construct GMP objects if they want to work with the numbers on a 32-bit platform.

If this class is constructed on a 64-bit integer platform, it's because someone unserialized some data from a 32-bit platform (as I do in tests). That's a very narrow edge case.

Reply:
Note: we agreed not to implement a cast handler, since that could result in overflow errors.